lib/rex/proto/http/client.rb

```diff
@@ -149,6 +149,39 @@ def request_cgi(opts = {})
     opts['ssl'] = self.ssl
     opts['ctype'] ||= 'application/x-www-form-urlencoded' if opts['method'] == 'POST'
 
+    if opts['form_data']
+      unless opts['form_data'].is_a?(::Array)
+        raise ::ArgumentError, "request_cgi: The provided `form_data` option is not valid. Expected: Array, Got: #{opts['form_data'].class}"
+      end
+
+      file_data = Rex::MIME::Message.new
+
+      opts['form_data'].each do |file_hash|
+        # The name of the HTTP form field
+        field_name = file_hash['name'].is_a?(::String) ? file_hash['name'] : nil
+
+        # Should we default to 'application/octet-stream', nil, or HttpClient's default of 'text/plain'?
+        # https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
+        mime_type = file_hash.fetch('mime_type', 'text/plain')
+
+        # Default to HttpClient's default option of '8bit'
+        encoding = file_hash.fetch('encoding', '8bit')
+
+        file_contents = get_file_data(file_hash['data'])
+
+        filename = file_hash.key?('filename') ? file_hash['filename'] : get_filename(file_hash['data']) || field_name
+
+        content_disposition = 'form-data'
+        content_disposition << "; name=\"#{field_name}\"" if field_name
+        content_disposition << "; filename=\"#{::CGI.escape(filename)}\"" if filename
+
+        file_data.add_part(file_contents, mime_type, encoding, content_disposition)
+      end
+
+      opts['data'] ? opts['data'] << file_data.to_s : opts['data'] = file_data.to_s
```

Comment:
Careful `<<` will mutate the string passed in as a parameter, this could introduce very subtle bugs 👀

Reply:
Respective code has been removed entirely, marking resolved.